src/Polysemy/Final/MTL.hs

```diff
@@ -24,64 +24,64 @@ import Polysemy.Writer hiding (tell, listen, pass)
 --
 -- /Beware/: Effects that aren't interpreted in terms of the final
 -- monad will have local state semantics in regards to 'Error' effects
--- interpreted this way. See 'interpretFinal'.
-runErrorFinal :: (Member (Final m) r, MonadError e m)
-              => Sem (Error e ': r) a
-              -> Sem r a
-runErrorFinal = interpretFinal $ \case
+-- interpreted this way. See 'Final'.
+errorToFinal :: (Member (Final m) r, MonadError e m)
+             => Sem (Error e ': r) a
+             -> Sem r a
+errorToFinal = interpretFinal $ \case
   Throw e   -> pure $ throwError e
   Catch m h -> do
     m' <- runS m
     h' <- bindS h
-    s <- getInitialStateS
+    s  <- getInitialStateS
     pure $ m' `catchError` (h' . (<$ s))
-{-# INLINE runErrorFinal #-}
+{-# INLINE errorToFinal #-}
 
 -----------------------------------------------------------------------------
 -- | Run a 'Reader' effect through a final 'MonadReader'
 --
 -- /Beware/: Effects that aren't interpreted in terms of the final
 -- monad will have local state semantics in regards to 'Reader' effects
--- interpreted this way. See 'interpretFinal'.
-runReaderFinal :: (Member (Final m) r, MonadReader i m)
-               => Sem (Reader i ': r) a
-               -> Sem r a
-runReaderFinal = interpretFinal $ \case
+-- interpreted this way. See 'Final'.
+readerToFinal :: (Member (Final m) r, MonadReader i m)
+              => Sem (Reader i ': r) a
+              -> Sem r a
+readerToFinal = interpretFinal $ \case
   Ask       -> liftS ask
   Local f m -> do
     m' <- runS m
     pure $ local f m'
-{-# INLINE runReaderFinal #-}
+{-# INLINE readerToFinal #-}
 
 -----------------------------------------------------------------------------
--- | Run a 'State' effect through a 'MonadState'
+-- | Run a 'State' effect in terms of an underlying 'MonadState' instance
 --
--- Despite the name, the target monad need not actually be the final
--- monad. The "-Final" suffix reflects that this interpreter
--- has the unusual semantics of interpreters that runs
+-- Although this is not a 'Final' inrepreeter, as the target monad need not
+-- actually be the final monad, 'stateToEmbed' still possesses the
+-- unusual semantics of interpreters that runs
 -- effects by embedding them into another monad.
 --
--- /Beware/: Effects that aren't interpreted in terms of the final
+-- /Beware/: Effects that aren't interpreted in terms of the embedded
 -- monad will have local state semantics in regards to 'State' effects
--- interpreted this way. See 'interpretFinal'.
-runStateFinal :: (Member (Embed m) r, MonadState s m)
-               => Sem (State s ': r) a
-               -> Sem r a
-runStateFinal = interpret $ \case
+-- interpreted this way. See 'Final'.
+stateToEmbed :: (Member (Embed m) r, MonadState s m)
```

Comment:
How well does this type get inferred? Things like this would always throw `freer-simple` for a loop, but maybe the plugin helps in this case?

Reply:
The plugin is awesome enough to resolve this.

Reply:
Even at the call site?

Reply:
[At the very least, this call site.](https://github.com/KingoftheHomeless/polysemy-zoo/blob/6527992539da9a8ff2497ede3af7dd0fa2d4e691/test/FinalSpec.hs#L112)

Reply:
remarkable.